specs/makeStyles.md

```diff
@@ -0,0 +1,322 @@
+
+### Passing overrides to children using `className`
+
+To avoid the ambivalence of using `className` vs `styles` props when passing overrides to children based on their type, we will use the dictionary side effect described in `useCSS` approach above.
+Parent component passes its overrides to a child as a list of classes in `className` prop. The child, takes the input classes and references the global dictionary to get back the necessary CSS attribute names to be able to perform the merge - the merge is implemented in `ax()` function as described in #16411.
+
+#### Alternative 1: encode the CSS property name to the hash
+
+Instead of using the global dictionary side effect, we can use part of the hash in the classname to encode the CSS property:
+![](makeStyles-classname.png)
+
+Then when merging in a child, we can merge the classes by replacing previous classnames with the same CSS property hash.
+This approach makes the classnames longer and according to the test does not make the merging any faster.
+
+#### Alternative 2: use special `styles` prop for overrides
+
+If the global dictionary does not perform as expected, we might return back to the previous v0 approach where we use a dedicated `styles` prop to pass overrides from parent to child.
+
+### Another libraries using similar approach
+
+This has been inspired by [Facebook stylex talk](https://www.youtube.com/watch?v=9JZHodNR184).
+Similar approach is also used in Attlasian [compiled v5](https://github.com/atlassian-labs/compiled).
+
+## Sample Code
+
+```jsx
+// Button.tsx
+const useButtonStyles = makeStyles([
+  [null, { cursor: 'pointer' }], // always applied
+  [props => props.primary, tokens => ({ backgroundColor: tokens.colors.primary.background })],
+  [props => pros.disabled, { backgroundColor: 'gray' }],
+]);
+
+const Button = ({ className, ...props }) => {
+  const buttonClasses = useButtonStyles(props); // button styles
+  const mergedClasses = ax(buttonClasses, className); // merge with `className` = overrides from parent
+  return <button className={mergedClasses} {...props} />;
+};
+
+// ButtonUsage.tsx
+const useButtonOverrides = makeStyles([[null, { backgroundColor: 'red' }]]);
+
+const ButtonUsage = () => {
+  const buttonOverrides = useButtonOverrides({});
+
+  return <Button classname={buttonOverrides}>I have always red background</Button>;
+};
+```
+
+## Issues, open questions
+
+### Theme-level style overrides
+
+As described in _Applying theme to styles_, a theme can only define tokens but can never change styles. There is a concern that this is not flexible enough and there might be a valid scenario where theme-specific style overrides will be required.
+How would that be implemented.
+Example API:
+
+```js
+// The Button styles need an id to pull overrides
+// from the theme:
+const makeStyles({ id: 'Button', { ... } });
```

Comment:
**q:** this is some different syntax on purpose ? (no array with tuples)

Reply:
It's not a real code, it only highlights an idea for a solution where we need to have theme specific overrides. Currently only this thing is possible:

```tsx
const useStyles = makeStyles([
  [null, { color: "black", borderWidth: "2px" }],
  [{ isDarkTheme: true }, { color: "white", borderWidth: "0" }] // should be always dark and have no border
]);
//
useStyles({ isDarkTheme: false /* some condition based on React Context */ });
```

_The main question is why not use tokens for this use case?_ Sometimes for overrides there are no proper tokens. We don't know if it's a pattern, that's why it stays as a proposal.

Reply:
Can this extended example be added to the doc as well ?

Reply:
added to `Sample Code` section